command: Add process_group method to Command struct

HarveyHunt opened the pull request:
Rust 1.64 stabilised the `process_group` method of the `std::os::unix::process::CommandExt` trait. This allows the caller to set the process group that a command is spawned in.

Expose that functionality from Tokio's `Command` structure.

This fixes #4312.

---

I'm not really familiar with the process for adding a new Tokio method, so I've gated it behind the `tokio_unstable` check - please let me know if I should be doing it differently. Thanks!

HarveyHunt commented:
It looks like the FreeBSD doc tests are failing because the nightly version used there is old, whilst the other FreeBSD tests are using stable.

Nightly is `1.61.0` ([logs](https://cirrus-ci.com/task/5346866794594304?logs=setup#L55))
Stable is `1.64.0` ([logs](https://cirrus-ci.com/task/6472766701436928?logs=setup#L55))

The Cirrus CI config is hardcoding an old nightly version - perhaps that could be updated?

https://github.com/tokio-rs/tokio/blob/master/.cirrus.yml#L5

ipetkov commented:
Doc tests are also failing, I would be okay to remove that example as it's pretty simplistic anyway

Darksonn commented on tokio/src/process/mod.rs:
```diff
@@ -690,6 +690,28 @@ impl Command {
         self
     }
 
+    /// Sets the process group ID (PGID) of the child process. Equivalent to a
+    /// setpgid call in the child process, but may be more efficient.
+    ///
+    /// Process groups determine which processes receive signals.
```
This should document that it is unstable, how to enable unstable features (a link is fine - see doc for other unstable methods), and that it will become stable when our MSRV is new enough to allow this in a stable manner. It would also be nice to mention that you can do this today by constructing an `std::process::Command` and using our `From` impl to convert it.

Darksonn reviewed (commented):
I don't mind adding this. Bumping the nightly version in CI is perfectly fine.

carllerche commented:
I'm fine adding it as an unstable feature w/ a comment saying to "stabilize" it when possible.

HarveyHunt commented:
Thanks for the feedback, I've updated the original commit to include some more documentation.

I've also added another commit that bumps the Cirrus CI version - please let me know if you'd prefer that in a separate PR.

Darksonn commented on tokio/src/process/mod.rs:
```diff
@@ -690,6 +690,38 @@ impl Command {
+    ///
+    /// If you want similar behaviour without using this unstable feature you can
+    /// create a [`std::process::Command`] and convert that into a
+    /// [`crate::process::Command`] using the `From` trait.
```
We don't use `crate` in our documentation.
```suggestion
    /// [`tokio::process::Command`] using the `From` trait.
```
If this link doesn't work out of the box, then you can add
```Rust
[`tokio::process::Command`]: crate::process::Command
```

HarveyHunt replied:
Ah, thanks for the pointer about the alias - I tried `tokio::process::Command` earlier and the docs wouldn't build. With that alias, it works. :)

Darksonn commented:
> I've also added another commit that bumps the Cirrus CI version - please let me know if you'd prefer that in a separate PR.

If it also requires a bunch of unrelated fixes of warnings, then I'd prefer a separate PR. Otherwise it is fine here.

Darksonn reviewed (approved):
Lgtm

Darksonn commented:
You'll need to fix the CI failure. Otherwise, I'm ready to merge this.

HarveyHunt commented:
Thanks for accepting! I've fixed up the warning in the doc test so this change should be good to merge now. :)

Darksonn reviewed (approved):
Thanks.

cablehead commented:
Hi, just checking if this feature still needs to be gated behind `tokio_unstable`. Thank you for all your work!

niklasf commented:
Yes. The [standard library function](https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.groups) was added in Rust 1.64.0, but tokio is at MSRV 1.63.
